h2o-algos/src/main/java/hex/glm/ComputationState.java

```diff
@@ -1026,7 +1026,7 @@ protected GramXY computeNewGram(DataInfo activeData, double [] beta, GLMParamete
     int [] activeCols = activeData.activeCols();
     int [] zeros = gt._gram.findZeroCols();
     GramXY res;
-    if(_parms._family != Family.multinomial && zeros.length > 0) {
+    if(_parms._family != Family.multinomial && zeros.length > 0 && zeros.length < activeData.activeCols().length) {
```

Comment:
Prevents `ArrayIndexOutOfBounds` in `Gram.dropCols`. Sometimes it would remove all the columns. IIRC some of the runs then tend to converge (I plan to verify if this claim still applies) but some end up with `NonSPDMatrixException`.

Reply:
If you uncomment the `# doTest("Custom test used for debugging divergence", test_custom)` in the test it should produce the error (tested on commit 50e0222651). If it doesn't, maybe the R tests are not as reproducible as I though so I persisted the data frame: [wrong_df.csv](https://github.com/h2oai/h2o-3/files/10509696/wrong_df.csv)

The stacktrace:
```java
   398 java.lang.ArrayIndexOutOfBoundsException: -1
   399         at hex.gram.Gram.dropCols(Gram.java:364)
   400         at hex.glm.ComputationState.computeNewGram(ComputationState.java:1030)
   401         at hex.glm.ComputationState.computeGram(ComputationState.java:1064)
   402         at hex.glm.GLM$GLMDriver.fitIRLSMML(GLM.java:2156)
   403         at hex.glm.GLM$GLMDriver.fitModel(GLM.java:2539)
   404         at hex.glm.GLM$GLMDriver.computeSubmodel(GLM.java:3022)
   405         at hex.glm.GLM$GLMDriver.doCompute(GLM.java:3163)
   406         at hex.glm.GLM$GLMDriver.computeImpl(GLM.java:3057)
   407         at hex.ModelBuilder$Driver.compute2(ModelBuilder.java:252)
   408         at hex.glm.GLM$GLMDriver.compute2(GLM.java:1508)
   409         at water.H2O$H2OCountedCompleter.compute(H2O.java:1677)
   410         at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)
   411         at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)
   412         at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:976)
   413         at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1479)
   414         at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)
```

Reply:
Now that I changed the `betaCnd` back to `_state.beta()` this dataset seems to work, so I'll have to find a new one to reproduce this issue.

Reply:
To reproduce it the following should work with the attached data.frame:
```r
df <- read.csv("95-1477-8.csv")

summary(MASS::glm.nb(result~., df))


hdf <- as.h2o(df)
m <- h2o.glm(y = "result", training_frame = hdf, family = "negativebinomial", link = "log",
             dispersion_parameter_method = "ml", standardize = FALSE, seed = 95)
```

[95-1477-8.csv](https://github.com/h2oai/h2o-3/files/10522369/95-1477-8.csv)